site/src/job_queue.rs

```diff
@@ -24,19 +27,118 @@ async fn enqueue_master_commits(ctxt: &Arc<SiteCtxt>) {
                 &master_commit.parent_sha,
                 pr,
                 master_commit.time,
-                BenchmarkRequestStatus::WaitingForParent,
+                BenchmarkRequestStatus::ArtifactsReady,
                 "",
                 "",
             );
             conn.insert_benchmark_request(&benchmark).await;
         }
     }
+    Ok(())
+}
+
+/// This function requires the sorts the priority order of the queue. It assumes
+/// that the only status is `ArtifactsReady`
+fn sort_benchmark_requests<'a>(
+    done: &mut HashSet<String>,
+    unordered_queue: &'a mut [BenchmarkRequest],
+) -> &'a [BenchmarkRequest] {
+    // A topological sort, where each "level" is additionally altered such that
+    // try commits come first, and then sorted by PR # (as a rough heuristic for
+    // earlier requests).
+
+    // Ensure the queue is ordered by status and the `commit_type`
+    unordered_queue.sort_unstable_by_key(|bmr| (bmr.status.rank(), bmr.commit_type.rank()));
+
+    let mut finished = 0;
+    while finished < unordered_queue.len() {
+        // The next level is those elements in the unordered queue which
+        // are ready to be benchmarked (i.e., those with parent in done or no
+        // parent).
+        let level_len = partition_in_place(unordered_queue[finished..].iter_mut(), |bmr| {
+            bmr.parent_sha().is_none_or(|parent| done.contains(parent))
```

Comment:
I would `expect()` the `parent_sha()` here, if it is missing, we have a bug in code.

Reply:
We shouldn't ever encounter release requests here, right? These should be sorted outside, and put in front of all master/try requests.

We need to filter for them though, because they will also have the `ArtifactsReady` status.

Reply:
Sorry, I again missed one invariant.. :laughing: In this function we should only be dealing with `ArtifactsReady` artifacts, these will always have a parent SHA. Try builds that are `WaitingForArtifacts` shouldn't ever come into this function.